pkg/exemplars/proxy.go

```diff
@@ -85,28 +86,21 @@ func (s *Proxy) Exemplars(req *exemplarspb.ExemplarsRequest, srv exemplarspb.Exe
 	for _, st := range s.exemplars() {
 		queryParts = queryParts[:0]
 
-	Matchers:
-		for _, matchers := range selectors {
-			matcherSet := make(map[string]struct{})
-			for _, m := range matchers {
-				for _, ls := range st.LabelSets {
-					if lv := ls.Get(m.Name); lv != "" {
-						if !m.Matches(lv) {
-							continue Matchers
-						} else {
-							// If the current matcher matches one external label,
-							// we don't add it to the current metric selector
-							// as Prometheus' Exemplars API cannot handle external labels.
-							continue
-						}
-					}
-					matcherSet[m.String()] = struct{}{}
-				}
+		for _, matcherSet := range selectors {
+			extLbls := st.LabelSets
+			if !store.LabelSetsMatch(matcherSet, extLbls...) {
+				continue
 			}
 
 			labelMatchers = labelMatchers[:0]
-			for m := range matcherSet {
-				labelMatchers = append(labelMatchers, m)
+			for _, m := range matcherSet {
+				if isExternalLabel(m.Name, extLbls) {
```

Comment:
I wonder if we should propagate labels matchers here, but check them in `tsdb.go`. Otherwise all tenants in a receiver will match whenever a single tenant matches.

Reply:
I guess it will fail when proxying via sidecar? Because we are also passing the query as is there. We could leave this as a follow up and maintain behavior for now. 

I think the best approach is to cover on all places, but would require ppl upgrading all components if they are running on distributed mode. Otherwise exemplars queries will not match anything.